### PR TITLE
Task 8 cards backend

### PR DESCRIPTION
**Pull request by alexanderjhughes**
I know, way to many files changed. This is why I am merging before adding rewards or any other functionality

### CODE REVIEW

File: backend/src/locales/es/userLocales.json [hunk 1]
```diff
@@ -3,8 +3,26 @@
     "userNotFound": "¡El usuario no existe!",
```
Comment (alexanderjhughes) [inline backend/src/locales/es/userLocales.json:3]: @lecler17 please review the Spanish translations I made. Hopefully they good :)

File: backend/src/models/BankModel.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      const sql = 'INSERT INTO rp_banks (bank_name, abbr) VALUES ($1, $2) RETURNING *';
+      const values = [bank.bank_name, bank.abbr.toLowerCase()];
+      const result = await client.query(sql, values);
+      if (!result.rows.length) {
+        throw new Error('error.bankNotCreated');
+      }
+      client.release();
+      return result.rows[0];
+    } catch (err: any) {
+      console.log('DB Error', err);
```
Comment (Tibesnoff) [inline backend/src/models/BankModel.ts:24]: No need for console log

File: backend/src/models/BankModel.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      const values = [bank_id];
+      const result = await client.query(sql, values);
+      client.release();
+      if (result.rows.length) {
+        return result.rows[0];
+      } else {
+        return null;
+      }
+    } catch (err: any) {
+      console.log('DB Error', err);
```
Comment (Tibesnoff) [inline backend/src/models/BankModel.ts:43]: Same here

File: backend/src/models/BankModel.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      const values = [bank_name];
+      const result = await client.query(sql, values);
+      client.release();
+      if (result.rows.length) {
+        return result.rows[0];
+      } else {
+        return null;
+      }
+    } catch (err: any) {
+      console.log('DB Error', err);
```
Comment (Tibesnoff) [inline backend/src/models/BankModel.ts:62]: Console log

File: backend/src/models/BankModel.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      const sql = 'SELECT * FROM rp_banks';
+      const result = await client.query(sql);
+      client.release();
+      if (result.rows.length) {
+        return result.rows;
+      } else {
+        return [];
+      }
+    } catch (err: any) {
+      console.log('DB Error', err);
```
Comment (Tibesnoff) [inline backend/src/models/BankModel.ts:80]: Honestly just leave them

File: backend/src/models/CardModel.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+export class CardModel {
+
+  async create(card: Card): Promise<Card> {
+    try {
+      const client = await dbPool.connect();
+      const cardCheck = await this.getByBin(card.card_bin);
+      if (cardCheck) {
+        throw new Error('error.cardFound');
+      }
+      const sql = 'INSERT INTO rp_cards (card_bin, card_name, card_brand_id, card_bank_id, card_type, card_level, card_country) VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *';
```
Comment (Tibesnoff) [inline backend/src/models/CardModel.ts:15]: This is more just an explanation. What does the dollar sign stuff do

File: backend/src/models/CardModel.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    try {
+      const client = await dbPool.connect();
+      const sql = 'SELECT * FROM rp_cards WHERE card_bin = $1';
+      const values = [card_bin];
+      const result = await client.query(sql, values);
+      client.release();
+      if (result.rows.length) {
+        return result.rows[0];
+      } else {
+        // check api, if in api then create card and return it here, if not return null
```
Comment (Tibesnoff) [inline backend/src/models/CardModel.ts:59]: Is there supposed to be code here?

File: backend/src/models/CardModel.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      console.log('DB Error', err);
+      const cardFriendlyError = i18n.t([err.message, 'error.default']);
+      throw new Error(cardFriendlyError);
+    }
+  }
+
+}
+
+export const CardModelInstance = new CardModel();
+export default CardModelInstance;
```
Comment (Tibesnoff) [inline backend/src/models/CardModel.ts:115]: I am slightly confused why there are multiple exports of the same thing

File: backend/src/types/bankTypes.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+export interface Bank {
```
Comment (Tibesnoff) [inline backend/src/types/bankTypes.ts:1]: I am voting for only needing an id and bank_name as well as not sending the id to the front end

File: package-lock.json
Comment (Tibesnoff) [inline package-lock.json:1]: Your package-lock sucks

**Review by Tibesnoff — COMMENTED** [review]
Looks good

File: backend/src/controllers/CardController.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  async createCard(req: Request, res: Response) {
+    const response: IJsonResponse = {
+      message: 'TLX API - Register Card',
+      success: true,
+      data: {}
+    };
+    const cardData = req.body as Card;
+    if (!cardData.card_bin || !cardData.card_bank_id || !cardData.card_brand_id || !cardData.card_type || !cardData.card_type || !cardData.card_level) {
```
Comment (Tibesnoff) [inline backend/src/controllers/CardController.ts]: I didn't see this before. You have card_type twice. This is causing an error

**Review by Tibesnoff — COMMENTED** [review]
error

File: backend/src/models/BankModel.ts (hunk 2 above)
Comment (alexanderjhughes) [reply backend/src/models/BankModel.ts:24]: This is needed for logging on the server

File: backend/src/models/BankModel.ts (hunk 3 above)
Comment (alexanderjhughes) [reply backend/src/models/BankModel.ts:43]: This is needed for logging on the server

File: backend/src/models/BankModel.ts (hunk 4 above)
Comment (alexanderjhughes) [reply backend/src/models/BankModel.ts:62]: This is needed for logging on the server

File: backend/src/models/CardModel.ts (hunk 6 above)
Comment (alexanderjhughes) [reply backend/src/models/CardModel.ts:15]: They represent the values we input

File: backend/src/types/bankTypes.ts (hunk 9 above)
Comment (alexanderjhughes) [reply backend/src/types/bankTypes.ts:1]: You need to send the id to the front end, that's how you identify the banks. Whenever you create a card you send the bank_id, etc etc. Abbr is going to be changed but we will still have it.

File: backend/src/models/CardModel.ts (hunk 8 above)
Comment (alexanderjhughes) [reply backend/src/models/CardModel.ts:115]: Just not supposed to be an export in front of the const, not that big of a deal

File: backend/src/models/CardModel.ts (hunk 7 above)
Comment (alexanderjhughes) [reply backend/src/models/CardModel.ts:59]: Not in this task

File: backend/src/controllers/CardController.ts (hunk 10 above)
Comment (alexanderjhughes) [reply backend/src/controllers/CardController.ts]: It doesn't cause any errors... removed it though

**Pull request merged**